### PR TITLE
Back off idle timers so the menu-bar app stops waking the main thread when idle

### PR DESCRIPTION
**Pull request by FuJacob**
## Summary

Two always-on timers woke the main thread continuously regardless of whether the user was typing. On Apple Silicon, idle package power is dominated by wake *frequency* rather than per-wake work, so these are a real idle-energy floor (Tier C of the #661 energy audit).

- **PermissionManager** polled `AXIsProcessTrusted` + two `CGPreflight*` calls every 2s for the entire session, even after every permission was already granted (the steady state for any established user). It now refreshes on `NSApplication.didBecomeActive` and keeps the 2s catch-up poll alive only while a *required* permission is still missing (onboarding). Once the required set is granted, the timer is torn down: zero idle wakeups thereafter.
- **FocusTracker** ran a fixed 50ms (20Hz) timer and used `FocusPollBackoff` to *skip* the expensive AX walk when idle, but the timer still fired 20x/s and hopped to the main actor only to no-op. The backoff now drives the timer *interval* itself (`base * stride`), so an idle machine wakes ~2x/s instead of 20x/s.

## Validation

```
xcodebuild -project Cotabby.xcodeproj -scheme Cotabby -destination 'platform=macOS' \
  -derivedDataPath build/DerivedData -onlyUsePackageVersionsFromResolvedFile \
  -only-testing:CotabbyTests/FocusPollBackoffTests CODE_SIGNING_ALLOWED=NO CODE_SIGNING_REQUIRED=NO test
# ** TEST SUCCEEDED **  Executed 9 tests, with 0 failures

swiftlint lint --quiet <changed files>
# exit 0
```

Built the full app + test target. Not yet exercised at runtime; behavior to confirm on next launch: with permissions already granted, no permission poll timer arms (and revoking a permission in System Settings is still reflected after returning to the app); focus tracking still follows the caret while typing and snaps back to fast cadence after an idle pause.

## Linked issues

Refs #661 (Tier C: idle timers).

## Risk / rollout notes

- **Permission revocation latency.** After the required set is granted there is no standing poll, so a revocation made in System Settings is reflected when the user next returns to the app (`didBecomeActive`) or opens any Cotabby surface (the menu, Settings, the Permissions pane, and onboarding all already call `refresh()` on appear). The app cannot function without those permissions anyway, and AX calls begin failing immediately on revocation, so the cache cannot drive harmful behavior in the gap. During onboarding (required not yet granted) the 2s poll still runs for snappy feedback.
- **Focus capture cadence is unchanged.** The timer now fires every `base * stride` and captures on every fire, which is the exact same spacing between AX walks as the previous "fire every base, skip to every stride" design. Only the no-op in-between wakes are removed. The timer is re-armed only when the stride actually changes, so active typing (stride 1) adds zero timer churn, and `refreshNow()` (called per keystroke) still resets to the base interval instantly.
- **API change.** `FocusPollBackoff.shouldCaptureOnTick()` is replaced by a `captureStride` multiplier. The only caller was `FocusTracker`; the unit tests are updated to assert the stride schedule and state machine directly (9 tests, all passing).
- No schema, settings, or pbxproj migration. No new files, so no `xcodegen generate` needed.

<h3>Greptile Summary</h3>

This PR reduces idle main-thread wakeups in two subsystems: `PermissionManager` now refreshes on `NSApplication.didBecomeActive` and only polls while a required permission is still missing (zero standing wakeups for established users), and `FocusTracker` drives its timer interval directly from the idle-backoff stride so the main thread wakes every `base × stride` instead of every `base` tick.

- **`FocusPollBackoff`** drops `ticksSinceCapture` / `shouldCaptureOnTick()` in favour of a `captureStride` property; `FocusTracker` multiplies the base interval by this stride and re-arms the timer only when the stride actually changes, keeping active typing at full cadence with zero timer churn.
- **`PermissionManager`** wires up a `didBecomeActive` observer, tears down the 2-second poll once `requiredPermissionsGranted` is true, and addresses the prior review's `@MainActor` isolation concern with `MainActor.assumeIsolated` in the activation callback.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the idle-backoff and permission-polling changes are well-scoped, `@MainActor` isolation is consistent throughout both modified actors, and the state machines are directly exercised by 9 passing unit tests.

Both behavioural changes preserve the original capture cadence (active users see no difference) and handle edge cases explicitly: permission revocation is reflected on the next activation or UI surface appear; timer re-arming is guarded by an interval-equality check so active typing causes zero timer churn. The one remaining inconsistency is a style/concurrency-annotation gap with no runtime consequence under current Swift settings.

All files look solid; `PermissionManager.swift` has the minor `DispatchQueue.main.async` inconsistency noted in the inline comment.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| Cotabby/Services/Focus/FocusTracker.swift | Replaces tick-skipping backoff with timer-interval backoff: adds `scheduledInterval` tracking, `scheduleTimer()`, and `rescheduleTimerIfIntervalChanged()`. Logic is sound — `@MainActor` isolation prevents races and `stop()` keeps `timer`/`scheduledInterval` in sync. |
| Cotabby/Services/Permission/PermissionManager.swift | Adds activation-based refresh and conditional polling — correct design. One minor inconsistency: the activation observer uses `MainActor.assumeIsolated` but the preserved timer callback in `startPollingIfNeeded` still uses `DispatchQueue.main.async`. |
| Cotabby/Support/FocusPollBackoff.swift | Simplifies the struct by removing `ticksSinceCapture` and replacing `shouldCaptureOnTick()` with a `captureStride` computed property. Logic and state machine are unchanged. |
| CotabbyTests/FocusPollBackoffTests.swift | Tests updated to match the new stride-based API; coverage is thorough — stride schedule, monotonicity, snap-back after idle, and reset all tested directly. |

</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `Cotabby/Services/Focus/FocusTracker.swift`, line 91-95 ([link](https://github.com/fujacob/cotabby/blob/46da6cecf9e040d27dcb4e08abbf7978d20ca774/Cotabby/Services/Focus/FocusTracker.swift#L91-L95)) 

   <a href="#">P2</a> Undocumented behavior change in `start()`: calling it while the timer is already running previously was a no-op; it now calls `refreshNow()`, which resets idle backoff (snap back to stride 1) and performs an immediate AX tree walk. If any callsite invokes `start()` while polling is already active — e.g. as a "ensure tracking is on" guard — this introduces an unintended backoff reset and an extra synchronous AX read that can briefly spike CPU. Consider documenting the new "restart capture and reset backoff" semantic in the method's doc comment.

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22fujacob%2Fcotabby%22%20on%20the%20existing%20branch%20%22energy-idle-timers%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22energy-idle-timers%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Cotabby%2FServices%2FFocus%2FFocusTracker.swift%0ALine%3A%2091-95%0A%0AComment%3A%0AUndocumented%20behavior%20change%20in%20%60start%28%29%60%3A%20calling%20it%20while%20the%20timer%20is%20already%20running%20previously%20was%20a%20no-op%3B%20it%20now%20calls%20%60refreshNow%28%29%60%2C%20which%20resets%20idle%20backoff%20%28snap%20back%20to%20stride%201%29%20and%20performs%20an%20immediate%20AX%20tree%20walk.%20If%20any%20callsite%20invokes%20%60start%28%29%60%20while%20polling%20is%20already%20active%20%E2%80%94%20e.g.%20as%20a%20%22ensure%20tracking%20is%20on%22%20guard%20%E2%80%94%20this%20introduces%20an%20unintended%20backoff%20reset%20and%20an%20extra%20synchronous%20AX%20read%20that%20can%20briefly%20spike%20CPU.%20Consider%20documenting%20the%20new%20%22restart%20capture%20and%20reset%20backoff%22%20semantic%20in%20the%20method's%20doc%20comment.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=fujacob%2Fcotabby&pr=665&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Cotabby%2FServices%2FFocus%2FFocusTracker.swift%0ALine%3A%2091-95%0A%0AComment%3A%0AUndocumented%20behavior%20change%20in%20%60start%28%29%60%3A%20calling%20it%20while%20the%20timer%20is%20already%20running%20previously%20was%20a%20no-op%3B%20it%20now%20calls%20%60refreshNow%28%29%60%2C%20which%20resets%20idle%20backoff%20%28snap%20back%20to%20stride%201%29%20and%20performs%20an%20immediate%20AX%20tree%20walk.%20If%20any%20callsite%20invokes%20%60start%28%29%60%20while%20polling%20is%20already%20active%20%E2%80%94%20e.g.%20as%20a%20%22ensure%20tracking%20is%20on%22%20guard%20%E2%80%94%20this%20introduces%20an%20unintended%20backoff%20reset%20and%20an%20extra%20synchronous%20AX%20read%20that%20can%20briefly%20spike%20CPU.%20Consider%20documenting%20the%20new%20%22restart%20capture%20and%20reset%20backoff%22%20semantic%20in%20the%20method's%20doc%20comment.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=fujacob%2Fcotabby&pr=665&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a>
</details>

<a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22fujacob%2Fcotabby%22%20on%20the%20existing%20branch%20%22energy-idle-timers%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22energy-idle-timers%22.%0A%0AFix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0ACotabby%2FServices%2FPermission%2FPermissionManager.swift%3A99-101%0AThe%20%60activationObserver%60%20callback%20in%20%60init%60%20was%20just%20changed%20to%20use%20%60MainActor.assumeIsolated%60%20%28addressing%20the%20previous%20review%20thread%29%2C%20but%20the%20timer%20callback%20in%20%60startPollingIfNeeded%60%20still%20dispatches%20via%20%60DispatchQueue.main.async%20%7B%20self%3F.refresh%28%29%20%7D%60.%20The%20timer%20is%20added%20to%20%60RunLoop.main%60%2C%20so%20it%20already%20fires%20on%20the%20main%20thread%3B%20the%20inner%20async%20dispatch%20is%20redundant%2C%20and%20calling%20the%20%60%40MainActor%60-isolated%20%60refresh%28%29%60%20from%20a%20non-isolated%20%60DispatchQueue.main.async%60%20closure%20has%20the%20same%20strict-concurrency%20concern%20the%20PR%20just%20fixed%20elsewhere%20in%20this%20file.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20let%20pollTimer%20%3D%20Timer%28timeInterval%3A%202.0%2C%20repeats%3A%20true%29%20%7B%20%5Bweak%20self%5D%20_%20in%0A%20%20%20%20%20%20%20%20%20%20%20%20MainActor.assumeIsolated%20%7B%20self%3F.refresh%28%29%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%60%60%60%0A%0A&repo=fujacob%2Fcotabby&pr=665&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0ACotabby%2FServices%2FPermission%2FPermissionManager.swift%3A99-101%0AThe%20%60activationObserver%60%20callback%20in%20%60init%60%20was%20just%20changed%20to%20use%20%60MainActor.assumeIsolated%60%20%28addressing%20the%20previous%20review%20thread%29%2C%20but%20the%20timer%20callback%20in%20%60startPollingIfNeeded%60%20still%20dispatches%20via%20%60DispatchQueue.main.async%20%7B%20self%3F.refresh%28%29%20%7D%60.%20The%20timer%20is%20added%20to%20%60RunLoop.main%60%2C%20so%20it%20already%20fires%20on%20the%20main%20thread%3B%20the%20inner%20async%20dispatch%20is%20redundant%2C%20and%20calling%20the%20%60%40MainActor%60-isolated%20%60refresh%28%29%60%20from%20a%20non-isolated%20%60DispatchQueue.main.async%60%20closure%20has%20the%20same%20strict-concurrency%20concern%20the%20PR%20just%20fixed%20elsewhere%20in%20this%20file.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20let%20pollTimer%20%3D%20Timer%28timeInterval%3A%202.0%2C%20repeats%3A%20true%29%20%7B%20%5Bweak%20self%5D%20_%20in%0A%20%20%20%20%20%20%20%20%20%20%20%20MainActor.assumeIsolated%20%7B%20self%3F.refresh%28%29%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%60%60%60%0A%0A&repo=fujacob%2Fcotabby&pr=665&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a>

<sub>Reviews (2): Last reviewed commit: ["Make the activation observer&#39;s main-acto..."](https://github.com/fujacob/cotabby/commit/ac3cbfbb3af6539e4dc5d04a5129dda6a560b66a) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36888777)</sub>

### CODE REVIEW

File: Cotabby/Services/Permission/PermissionManager.swift [hunk 1]
```diff
@@ -17,22 +17,35 @@ final class PermissionManager: ObservableObject {
...
-        refresh()
-        let pollTimer = Timer(timeInterval: 2.0, repeats: true) { [weak self] _ in
-            DispatchQueue.main.async { self?.refresh() }
+        activationObserver = NotificationCenter.default.addObserver(
+            forName: NSApplication.didBecomeActiveNotification,
+            object: nil,
+            queue: .main
+        ) { [weak self] _ in
+            self?.refresh()
         }
```
[the remark this answers is not in the document]
Comment (FuJacob) [reply Cotabby/Services/Permission/PermissionManager.swift:45]: Fixed in ac3cbfb: wrapped the callback body in `MainActor.assumeIsolated`, matching the SystemMetricsStore/InputMonitor pattern. Kept it synchronous (vs `Task { @MainActor }`) so permission state is fresh before any same-turn UI reads on activation.
